sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -850,8 +851,9 @@ public boolean requiresDeduping() {
     private final String name;
     private Consumer<byte[], byte[]> consumer;
     private final List<PartitionState> partitionStates;
-    private KafkaRecord<K, V> curRecord;
-    private Instant curTimestamp;
+    // curRecord and curTimestamp are accessed outside advance(), which might be another thread.
+    private AtomicReference<KafkaRecord<K, V>> curRecord = new AtomicReference<>();
```

Comment:
This is extremely smelly; the current record and current timestamp must be updated atomically and simultaneously, so using two separate atomic references is not sufficient - you need to store the timestamped record here

Reply:
Yeah.. I thought of explicitly commenting on that explicitly. Right now we don't depend on them to be consistent. But I agree, simpler to do the right thing.

Reply:
Fixed. Please take look.